src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+
+namespace System.Threading
+{
+    public sealed class ReaderWriterLock : CriticalFinalizerObject
+    {
+        private const int InvalidThreadID = -1;
+        private const ushort MaxAcquireCount = ushort.MaxValue;
+        private static readonly int DefaultSpinCount = Environment.ProcessorCount != 1 ? 500 : 0;
+
+        private static long s_mostRecentLockID;
+
+        private ManualResetEvent _readerEvent;
+        private AutoResetEvent _writerEvent;
+        private long _lockID;
+        private volatile int _state;
+        private int _writerID = InvalidThreadID;
+        private int _writerSeqNum;
+        private ushort _writerLevel;
+
+#if RWLOCK_STATISTICS
+        private int _readerEntryCount;
+        private int _readerContentionCount;
+        private int _writerEntryCount;
+        private int _writerContentionCount;
+        private int _eventsReleasedCount;
+#endif // RWLOCK_STATISTICS
+
+        public ReaderWriterLock()
+        {
+            _lockID = Interlocked.Increment(ref s_mostRecentLockID);
+            GC.SuppressFinalize(this);
+        }
+
+        ~ReaderWriterLock()
+        {
+        }
+
+        public bool IsReaderLockHeld
+        {
+            get
+            {
+                LockEntry lockEntry = LockEntry.GetCurrent(_lockID);
+                if (lockEntry != null)
+                {
+                    return lockEntry._readerLevel != 0;
+                }
+                return false;
+            }
+        }
+
+        public bool IsWriterLockHeld
+        {
+            get
+            {
+                return _writerID == GetCurrentThreadID();
+            }
+        }
+
+        public int WriterSeqNum
+        {
+            get
+            {
+                return _writerSeqNum;
+            }
+        }
+
+        public bool AnyWritersSince(int seqNum)
+        {
+            if (_writerID == GetCurrentThreadID())
+            {
+                ++seqNum;
+            }
+            return (uint)_writerSeqNum > (uint)seqNum;
+        }
+
+        public void AcquireReaderLock(int millisecondsTimeout)
+        {
+            if (millisecondsTimeout < -1)
+            {
+                throw GetInvalidTimeoutException(nameof(millisecondsTimeout));
+            }
+
+            LockEntry lockEntry = LockEntry.GetOrCreateCurrent(_lockID);
+
+            // Check for the fast path
+            if (Interlocked.CompareExchange(ref _state, LockStates.Reader, 0) == 0)
+            {
+                Debug.Assert(lockEntry._readerLevel == 0);
+            }
+            // Check for nested reader
+            else if (lockEntry._readerLevel != 0)
+            {
+                Debug.Assert((_state & LockStates.ReadersMask) != 0);
+
+                if (lockEntry._readerLevel == MaxAcquireCount)
+                {
+                    throw new OverflowException(SR.Overflow_UInt16);
+                }
+                ++lockEntry._readerLevel;
```

Comment:
What makes this safe?  Is it that if we're here, we know we're a writer and thus we're the only thread that could be mutating this?